test/test_testing.py

```diff
@@ -1305,5 +1307,98 @@ def test_regression_info_new_job(self):
         )
 
 
+class TestFrameworkUtils(TestCase):
+    tests = [
+        'super_long_test',
+        'long_test1',
+        'long_test2',
+        'normal_test1',
+        'normal_test2',
+        'normal_test3',
+        'short_test1',
+        'short_test2',
+        'short_test3',
+        'short_test4',
+        'short_test5',
+    ]
+
+    test_times = {
+        'super_long_test': (55, 1),
+        'long_test1': (22, 2),
+        'long_test2': (18, 2),
+        'normal_test1': (9, 2),
+        'normal_test2': (7, 2),
+        'normal_test3': (5, 2),
+        'short_test1': (1, 2),
+        'short_test2': (0.6, 3),
+        'short_test3': (0.4, 5),
+        'short_test4': (0.3, 1),
+        'short_test5': (0.01, 2),
+    }
+
+    def test_calculate_2_shards_with_complete_test_times(self):
+        expected_shards = [
+            (60, ['super_long_test', 'normal_test3']),
+            (58.31, ['long_test1', 'long_test2', 'normal_test1', 'normal_test2', 'short_test1', 'short_test2',
+                     'short_test3', 'short_test4', 'short_test5'])
+        ]
+        self.assertEqual(expected_shards, calculate_shards(2, self.tests, self.test_times))
+
+
+    def test_calculate_5_shards_with_complete_test_times(self):
+        expected_shards = [
+            (55, ['super_long_test']),
+            (22, ['long_test1', ]),
+            (18, ['long_test2', ]),
+            (11.31, ['normal_test1', 'short_test1', 'short_test2', 'short_test3', 'short_test4', 'short_test5']),
+            (12, ['normal_test2', 'normal_test3']),
+        ]
+        self.assertEqual(expected_shards, calculate_shards(5, self.tests, self.test_times))
+
+
+    def test_calculate_2_shards_with_incomplete_test_times(self):
+        incomplete_test_times = {k: v for k, v in self.test_times.items() if 'test1' in k}
+        expected_shards = [
+            (22, ['long_test1', 'long_test2', 'normal_test3', 'short_test3', 'short_test5']),
+            (10, ['normal_test1', 'short_test1', 'super_long_test', 'normal_test2', 'short_test2', 'short_test4']),
+        ]
+        self.assertEqual(expected_shards, calculate_shards(2, self.tests, incomplete_test_times))
+
+
+    def test_calculate_5_shards_with_incomplete_test_times(self):
+        incomplete_test_times = {k: v for k, v in self.test_times.items() if 'test1' in k}
+        expected_shards = [
+            (22, ['long_test1', 'normal_test2', 'short_test5']),
+            (9, ['normal_test1', 'normal_test3']),
+            (1, ['short_test1', 'short_test2']),
+            (0, ['super_long_test', 'short_test3']),
+            (0, ['long_test2', 'short_test4']),
+        ]
+        self.assertEqual(expected_shards, calculate_shards(5, self.tests, incomplete_test_times))
+
+
+    def test_calculate_2_shards_against_optimal_shards(self):
+        for _ in range(100):
+            random_times = {k: (random.random() * 10, 1) for k in self.tests}
```

Comment:
Instead of using unseeded randomness (making this test nondeterministic and thus a bit harder to debug if/when it fails), perhaps consider using [Hypothesis](https://hypothesis.readthedocs.io/en/latest/)? (I believe we already require it for running the tests)

Reply:
oh wow didn't know hypothesis was just python's Quickchecks--real cool!

Reply:
or easier, you can also just set a seed for random https://docs.python.org/3/library/random.html

Reply:
Oops too late I already used hypothesis (which was quite fun, would recommend)